src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -0,0 +1,132 @@
+package io.iohk.ethereum.db.storage
+
+import akka.util.ByteString
+import io.iohk.ethereum.db.storage.NodeStorage.{NodeEncoded, NodeHash}
+import io.iohk.ethereum.mpt.NodesKeyValueStorage
+import io.iohk.ethereum.rlp.RLPImplicitConversions.{toRlpList, _}
+import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.rlp.{encode => rlpEncode, _}
+
+/**
+  * This class is used to store Nodes (defined in mpt/Node.scala), by using:
+  * Key: hash of the RLP encoded node
+  * Value: the RLP encoded node
+  */
+class ReferenceCountNodeStorage(nodeStorage: NodeStorage, blockNumber: Option[BigInt] = None) extends NodesKeyValueStorage {
+
+  import ReferenceCountNodeStorage._
+
+  override def get(key: ByteString): Option[NodeEncoded] = nodeStorage.get(key).map(_.toStoredNode).map(_.nodeEncoded.toArray)
+
+  override def update(toRemove: Seq[NodeHash], toUpsert: Seq[(NodeHash, NodeEncoded)]): NodesKeyValueStorage = {
+
+    require(blockNumber.isDefined)
+
+    val bn = blockNumber.get
+    val upsertChanges = toUpsert.foldLeft(Map.empty[NodeHash, StoredNode]) { (storedNodes, toUpsertItem) =>
+      val (nodeKey, nodeEncoded) = toUpsertItem
+      val storedNode: StoredNode = storedNodes.get(nodeKey)
```

Comment:
Yes it can. For example, in the [following test](https://github.com/input-output-hk/etc-client/blob/master/src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala#L436) (that was created before this test) the same node is being stored twice.

Maybe it's an issue with MPT implementation but I don't think it's a bug. will discuss it with @ntallar